contracts/discovery/GNS.sol

```diff
@@ -202,155 +173,157 @@ contract GNS is GNSV1Storage, GraphUpgradeable, IGNS {
+     * NOTE: This function updates the sequence ID for the account
+     * @return Sequence ID for the account
+     */
+    function _nextAccountSeqID(address _account) internal returns (uint256) {
+        uint256 seqID = nextAccountSeqID[_account];
+        nextAccountSeqID[_account] = nextAccountSeqID[_account].add(1);
+        return seqID;
     }
 
     /**
-     * @dev Private function used by both external publishing functions
-     * @param _graphAccount Account that is publishing the subgraph
-     * @param _subgraphNumber Subgraph number for the account
+     * @dev Publish a new version of an existing subgraph.
+     * @param _subgraphID Subgraph ID
      * @param _subgraphDeploymentID Subgraph deployment ID of the version, linked to the name
      * @param _versionMetadata IPFS hash for the subgraph version metadata
      */
-    function _publishVersion(
-        address _graphAccount,
-        uint256 _subgraphNumber,
+    function publishNewVersion(
+        uint256 _subgraphID,
         bytes32 _subgraphDeploymentID,
         bytes32 _versionMetadata
-    ) private {
+    ) external override notPaused onlySubgraphOwner(_subgraphID) {
+        // Subgraph must exist
+        require(
+            isPublished(_subgraphID),
+            "GNS: Cannot update version if not published, or has been deprecated"
+        );
+
+        // New subgraph deployment must be non-empty
         require(_subgraphDeploymentID != 0, "GNS: Cannot set deploymentID to 0 in publish");
 
-        // Stores a subgraph deployment ID, which indicates a version has been created
-        subgraphs[_graphAccount][_subgraphNumber] = _subgraphDeploymentID;
+        // Perform the upgrade from the current subgraph deployment to the new one.
+        // This involves burning all signal from the old deployment and using the funds to buy
+        // from the new deployment.
+        // This function will also make the change to target to the new deployment.
+        _upgradeSubgraph(_subgraphID, _subgraphDeploymentID);
 
-        // Emit version and name data
-        emit SubgraphPublished(
-            _graphAccount,
-            _subgraphNumber,
-            _subgraphDeploymentID,
-            _versionMetadata
-        );
+        emit SubgraphVersionUpdated(_subgraphID, _subgraphDeploymentID, _versionMetadata);
     }
 
     /**
-     * @dev Deprecate a subgraph. Can only be done by the graph account owner.
-     * @param _graphAccount Account that is deprecating the subgraph
-     * @param _subgraphNumber Subgraph number for the account
+     * @dev Deprecate a subgraph. The bonding curve is destroyed, the vSignal is burned, and the GNS
+     * contract holds the GRT from burning the vSignal, which all curators can withdraw manually.
+     * Can only be done by the subgraph owner.
+     * @param _subgraphID Subgraph ID
      */
-    function deprecateSubgraph(address _graphAccount, uint256 _subgraphNumber)
+    function deprecateSubgraph(uint256 _subgraphID)
```

Comment:
will we burn the NFT here?

Reply:
I was originally thinking about burning the NFT but I think we could just keep it there and be useless. The reason for that is that if we want to burn I need to get it transferred to the contract. I see no harm in the user keeping the NFT.

Reply:
Might be worth while to think through what the Front end experience will be like. 

If we are displaying NFTs, or someone goes to OpenSea and sees subgraph NFTs, it might be confusing. 

Also if a subgraph developer has like 15 subgraphs, 8 of them deprecated, they might need to sift through them, like on etherscan. 

 maybe we burn, maybe we dont. lets think about it

Reply:
I decided to burn the token so it's not sitting there as something that someone can transfer but with no purpose to manage the subgraph.

No need to transfer the NFT to the GNS to burn it, we can use the internal `_burn` function because the GNS is the NFTRegistry.

```
    function _burn(uint256 tokenId) internal virtual {
        address owner = ERC721Upgradeable.ownerOf(tokenId); // internal owner

        _beforeTokenTransfer(owner, address(0), tokenId);

        // Clear approvals
        _approve(address(0), tokenId);

        // Clear metadata (if any)
        if (bytes(_tokenURIs[tokenId]).length != 0) {
            delete _tokenURIs[tokenId];
        }

        _holderTokens[owner].remove(tokenId);

        _tokenOwners.remove(tokenId);

        emit Transfer(owner, address(0), tokenId);
    }
```